### PR TITLE
WT-9988 Upgrade the wt util bson parsing python script.

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: tools/wt_to_mdb_bson.py [hunk 1]
```diff
@@ -53,24 +143,67 @@ def decode_data_section(mdb_file_contents, data_index):
...
+    else:
+        print('Invalid mode specified.')
+        util_usage()
+
+    # Does the user plan on passing wt's location and a file?
+    if len(sys.argv) > 3:
+        if sys.argv[3] != '-i':
+            print('Invalid option specified.')
+            util_usage()
+        wt_output = execute_wt(mode, sys.argv[4], sys.argv[5])
```
Comment (ajmorton) [inline tools/wt_to_mdb_bson.py:196]: `./wt_to_mdb_bson -m printlog -f ./wt` doesn't require a .wt file so `sys.argv[5]` can be an out of bounds access here.

Given we already know `mode` I'd just add a
`wt_file = None if mode == Mode.PRINTLOG else sys.argv[5]` and pass that into `execute_wt`

File: tools/wt_to_mdb_bson.py [hunk 2]
```diff
@@ -26,23 +26,113 @@
...
+    print("Usage: wt_to_mdb_bson -m {dump|verify|printlog} [-f] [path_to_wt] [filename]")
+    print('\t-m the intended mode that the wt util operated in or will be executed using.')
+    print('\t-f the location of the wt util.')
+    sys.exit(1)
+
+# BSON printer helper.
+def print_bson(bson):
+    return pprint.pformat(bson, indent=1).replace('\n', '\n\t  ')
+
+# A utility function for converting verify byte output into parsible hex.
```
Comment (ajmorton) [inline tools/wt_to_mdb_bson.py:71]: Can we add a comment explaining the only change is to insert `\x` when we find a single `\` by itself?

File: tools/wt_to_mdb_bson.py [hunk 3]
```diff
@@ -53,24 +143,67 @@ def decode_data_section(mdb_file_contents, data_index):
...
+def execute_wt(mode, wtpath, filename):
+    if mode == Mode.DUMP:
+        return subprocess.check_output(
+            [wtpath, "-r", "dump", "-x", "file:" + filename], universal_newlines=True).splitlines()
+    elif mode == Mode.VERIFY:
+        return subprocess.check_output(
+            [wtpath, "-r", "verify", "-d", "dump_pages", "file:" + filename], universal_newlines=True).splitlines()
+    else:
+        return subprocess.check_output(
+            [wtpath, "-r", "printlog", "-u", "-x"], universal_newlines=True).splitlines()
```
Comment (dgottlieb) [inline tools/wt_to_mdb_bson.py:169]: This is a neat shorthand for having the script invoke `wt`. Couple remarks:
* I don't see this using any of the -C to set additional configuration. Specifically saying the WT log is kept in the `journal` directory. And that the log is compressed with snappy.
* I would expect some of these calls (particularly `printlog`) to have GBs of output. And I believe this is buffering all the output into a string before splitting. 20 years ago, python would choke on that growing a string to that size. Maybe things have changed.

File: tools/wt_to_mdb_bson.py [hunk 4]
```diff
@@ -26,23 +26,113 @@
...
+        ret += ch + ch
+        idx += 2
+    return codecs.escape_decode(ret)[0]
+
+# Converts the output of ./verify -d dump_pages to bson.
+def wt_verify_to_bson(wt_output):
+    pattern = re.compile('V {(.*?)}$')
+    for line in wt_output:
+        matches = pattern.findall(line.strip())
+        if not matches:
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:97]: That isn't functionally the same as you would print twice.

File: tools/wt_to_mdb_bson.py [hunk 5]
```diff
@@ -26,23 +26,113 @@
...
 def util_usage():
-    print("Usage: wt_to_mdb_bson <path_to_wt> filename")
+    print("Usage: wt_to_mdb_bson -m {dump|verify|printlog} [-f] [path_to_wt] [filename]")
+    print('\t-m the intended mode that the wt util operated in or will be executed using.')
+    print('\t-f the location of the wt util.')
+    sys.exit(1)
+
+# BSON printer helper.
+def print_bson(bson):
+    return pprint.pformat(bson, indent=1).replace('\n', '\n\t  ')
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:69]: Not sure this was borrowed from the original script, I'd prefer to keep it that way.

File: tools/wt_to_mdb_bson.py (hunk 2 above)
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:71]: Unfortunately I don't actually know the answer to that as it was also copied from an existing script. I will look into it.

File: tools/wt_to_mdb_bson.py (hunk 3 above)
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:169]: I added the -C configuration for printlog. I am hoping the many GBs issue isn't a big problem. If people report that issue we can look into a buffered solution.

File: tools/wt_to_mdb_bson.py (hunk 1 above)
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py]: 73875cd5e

File: tools/wt_to_mdb_bson.py [hunk 6]
```diff
@@ -26,23 +26,113 @@
...
+# Those tools each perform a different function, and their usages are varied as such the script
+# needs to handle their output separately. Originally many scripts existed for this purpose, the
+# intent of this script is to provide a single place to perform all bson conversions.
+#
+# This script takes input of two forms, either through stdin or the user can pass the wt util
+# location and the filename and the script will execute the required wt util command. When running
+# with -f the script must be executed in the same directory as the database.
+#
+# Some example usages are:
+#    - ./wt -r dump -x | ./wt_to_mdb_bson -m dump
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py]: 73875cd5e

File: tools/wt_to_mdb_bson.py [hunk 7]
```diff
@@ -26,23 +26,113 @@
...
+        idx += 2
+    return codecs.escape_decode(ret)[0]
+
+# Converts the output of ./verify -d dump_pages to bson.
+def wt_verify_to_bson(wt_output):
+    pattern = re.compile('V {(.*?)}$')
+    for line in wt_output:
+        matches = pattern.findall(line.strip())
+        if not matches:
+            print(line, end='')
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:98]: 73875cd5e

File: tools/wt_to_mdb_bson.py [hunk 8]
```diff
@@ -53,24 +143,67 @@ def decode_data_section(mdb_file_contents, data_index):
...
+        mode = Mode.VERIFY
+    elif mode_str == 'printlog':
+        mode = Mode.PRINTLOG
+    else:
+        print('Invalid mode specified.')
+        util_usage()
+
+    # Does the user plan on passing wt's location and a file?
+    if len(sys.argv) > 3:
+        if sys.argv[3] != '-i':
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py]: 73875cd5e

File: tools/wt_to_mdb_bson.py [hunk 9]
```diff
@@ -26,23 +26,113 @@
...
+
+        obj = bson.decode_all(convert_byte(matches[0]))[0]
+        print(line, end='')
+        print('\t  %s' % (print_bson(obj),))
+
+# Converts the output of ./wt printlog -x -u.
+# Doesn't convert hex keys as I don't think they're bson.
+def wt_printlog_to_bson(wt_output):
+    pattern_value = re.compile('value-hex\": \"(.*)\"')
+    pattern_key = re.compile('key-hex\": \"(.*)\"')
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py]: 73875cd5e

File: tools/wt_to_mdb_bson.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajmorton) [reply tools/wt_to_mdb_bson.py:97]: Isn't that expected? 
In the current code both the `if not matches:` and its implicit else branch are calling `print(line, end='')`. It's just when we have a match we also attach the decoded `print('\t  %s' % (print_bson(obj),))`

File: tools/wt_to_mdb_bson.py (hunk 2 above)
Comment (dgottlieb) [reply tools/wt_to_mdb_bson.py:71]: IIRC, @ajmorton has it right. C's "format this char* as a combination of ascii and escaped hex values" has no natural way to deserialize back into a python byte array.

File: tools/wt_to_mdb_bson.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:97]: When we do the `print(line, end'')` we continue and skip the decode print.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py:97]: Ah I see it now, thanks.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply tools/wt_to_mdb_bson.py]: af1554013

**Review by ajmorton — APPROVED** [review]
LGTM 👍

**Pull request merged**